### DISCUSSION TITLE
How can I properly instantiate models?

### DISCUSSION THREAD

**Discussion by etudor** [typegoose/typegoose#757] [Q&A]
I have the following model which is a typegraphql and a typegoose model  as well
```ts
@ObjectType()
class AssetMongo {
  @Field(() => ID)
  @prop({ required: true, es_type: 'keyword', es_indexed: true })
  title!: string
...
}
```
and this is the model class 
```ts
export const AssetModel = getModelForClass(AssetMongo)
```
How should I properly instantiate the AssetMongo so that It's correct in typegraphql and correctly typed by typescript.

- case 1
```ts
const asset = new AssetMongo()
asset.save() // will break
// in this case typescript doesn't sees the .save() property
```

- case 2
```ts
const asset = new AssetModel()
asset.save() // works
// this works, typescript sees the other properties but breaks typegraphql
```
- with this In TypeGraphql I have 2 options

```ts
@ObjectType()
class AssetRes {
  @Field(() => [AssetModel])
  assets: AssetModel[] 
}
```
- this will break typegraphql as AssetModel is not a Typegraphql object, AssetMongo is

```ts
@ObjectType()
class AssetRes {
  @Field(() => [AssetMongo])
  assets: AssetMongo[] 
}
```
- this breaks TS because `const asset = new AssetModel()` can't be assigned to type AssetMongo

How do you properly use Typegoose with Typegraphql?  
What is the correct typescript of the AssetMongo? `AssetMongo, DocumentType<AssetMongo>, ReturnType<AssetMongo>, AssetModel`? 

```ts
const asset: AssetMongo, DocumentType<AssetMongo>, ReturnType<AssetMongo>, AssetModel??  = AssetRepo.findByUuid(uuid)
```

**Comment by hasezoey** [comment]
> case 1
> // in this case typescript doesn't sees the .save() property

that is to be expected, document methods only exist on a document, and mongoose Models and Documents are not instances of a class

> case 2
> // this works, typescript sees the other properties but breaks typegraphql

i dont know much about (type)graphql so i cant help there, but maybe provide where / what does not work (if you dont mean the cases below in your question)

> this will break typegraphql as AssetModel is not a Typegraphql object, AssetMongo is
> this breaks TS because const asset = new AssetModel() can't be assigned to type AssetMongo

i dont quite know what you want to achieve there (and also from what little i know about graphql), you probably want `AssetMongo`(the class directly) (or by translating the Class type to a POJO using the mentioned functions [in this stackoverflow answer (the updated section)](https://stackoverflow.com/a/55479659/8944059) and also using some kind of transform) or if you actually want documents at that point (which is from my knowledge about graphql not quite correct) you can use typegoose's type [`DocumentType<YourClassNameHere>`](https://typegoose.github.io/typegoose/docs/api/types/document-type)

> `...AssetRepo.findByUuid...`

i dont know what this function is, so i cannot provide a accurate type, though if it is the same as `Model.findOne().exec()`, then the proper type would be `DocumentType<YourClass>`

---

if you dont know where to put which types where (about `DocumentType` and `ReturnModelType`), read through [the typegoose documentation](https://typegoose.github.io/typegoose/docs/guides/quick-start-guide)
